docs/specification/new-spec/consensus-reactor.md

```diff
@@ -0,0 +1,239 @@
+	ProposalBlockPartsHeader PartSetHeader 
+	ProposalBlockParts       BitArray       
+	ProposalPOLRound         int                 // Proposal's POL round. -1 if none.
+	ProposalPOL              BitArray            // nil until ProposalPOLMessage received.
+	Prevotes                 cmn.BitArray        // All votes peer has for this round
+	Precommits               cmn.BitArray        // All precommits peer has for this round
+	LastCommitRound          int                 // Round of commit for last height. -1 if none.
+	LastCommit               BitArray            // All commit precommits of commit for last height.
+	CatchupCommitRound       int                 // Round that we have commit for. Not necessarily unique. -1 if none.
+	CatchupCommit            BitArray            // All commit precommits peer has for this height & CatchupCommitRound
+}
+```   
+
+## Receive method of Consensus reactor
+
+When a message is received from a peer p, it is treated differently based on the channel message belongs to.
+
+TODO: Add high level description of the logic of Receive method. 
+
+## Gossip Data Routine
+
+It is used to send the following messages to the peer: `BlockPartMessage`, `ProposalMessage` and 
+`ProposalPOLMessage` on the DataChannel. The gossip data routine is based on the local RoundState (denoted rs) 
+and the known PeerRoundState (denotes prs). The routine repeats forever the logic shown below:
+
+```
+1a) if rs.ProposalBlockPartsHeader == prs.ProposalBlockPartsHeader and the peer does not have all the proposal parts then
+        Part = pick a random proposal block part the peer does not have 
+        Send BlockPartMessage(Height, Round, Part) to the peer on the DataChannel 
+        if send returns true, record that the peer knows the corresponding block Part
+	    Continue  
+ 
+1b) if (0 < prs.Height) and (prs.Height < rs.Height) then
+        help peer catch up using gossipDataForCatchup function
+        Continue
+
+1c) if (rs.Height != prs.Height) or (rs.Round != prs.Round) then 
+        Sleep PeerGossipSleepDuration
+        Continue   
+
+//  at this point rs.Height == prs.Height and rs.Round == prs.Round
+1d) if (rs.Proposal != nil and !prs.Proposal) then 
+        Send ProposalMessage(rs.Proposal) to the peer
+        if send returns true, record that the peer knows Proposal
+	    if 0 <= rs.Proposal.POLRound then
+	    polRound = rs.Proposal.POLRound 
+        prevotesBitArray = rs.Votes.Prevotes(polRound).BitArray() 
+        Send ProposalPOLMessage(rs.Height, polRound, prevotesBitArray)
+        Continue  
+
+2)  Sleep PeerGossipSleepDuration 
+```
+
+## Gossip Votes Routine
+
+It is used to send the following message: `VoteMessage` on the VoteChannel.
+The gossip votes routine is based on  the local RoundState (denoted rs) 
+and the known PeerRoundState (denotes prs). The routine repeats forever the logic shown below:
+
+```
+1a) if rs.Height == prs.Height then
```

Comment:
If our peer is lagging behind (it is in smaller round) why we send him messages from his round. Shouldn't we help him to catch up asap to the round where we are? Although we need to be sure that important messages (those with Polkas are being transmitted). Also, does it make sense sending Prevotes for prs.ProposalPOLRound if we are in higher round?

Reply:
If prs.Step == RoundStepNewHeight, it's actually in a waiting period before starting the new height's first round.  During this period, it still wants more straggler votes from the last commit (prs.Height-1), because when it's a proposer, it will want to include as many precommits in the proposal block's "lastCommit" as possible.

(It wants to include more precommits in the proposal block's lastCommit because it gets slightly more block rewards.  this is to encourage validators to invest in gossip network robustness.  If the peer wants to stop receiving lastCommit precommits, it can signal so via a NewRoundStep message of (H,0,RoundStepNewRound).  Some validators may tweak the "timeoutCommit" parameter to try to wait more or less, I suppose.  They would be trading off the risk of their proposal not getting out in time, vs the risk of losing out on precommit inclusion rewards.

Sidenote... If say Tendermint V2 allows all proposers to propose simultaneously with some randomness of some sort at height H, round 0, (and each node would only gossip the "most fit" proposal, so this need not hurt the propagation time of the winning proposal too much).

Reply:
I understand your point about straggler votes and it's valid comment. I was having something different on mind: if we are at round 5 and our peer is at round 3. Should we send him messages from round 3 although we failed to commit at that round (as we are in round 5)? If we are in round 5 this means we received 2f+1 (for simplicity lets assume n = 3f+1) votes from round 4 so we know at least f+1 correct processes also failed to commit at round 3 as they moved to round 4. If we send to the peer messages from round 3 we will need to wait (potentially) significant time before we have all correct processes in the same round, so we can decide and commit a block. With protocols where there is full connectivity between processes, we normally have catchup mechanisms where a process jump to higher round if it sees f+1 messages from higher round. This guarantees that after GST we have bounded (relatively short) time period after which all correct processes are in the same round. It's more tricky to ensure such property with p2p communication as we need to be careful with DDoS attacks.